sm/agent/src/agent.rs

```diff
@@ -0,0 +1,326 @@
+        let request_topics =
+            TopicFilter::new("tedge/commands/req/software/#").expect("Invalid topic");
+
+        let request_topic_list =
+            Topic::new(SoftwareListRequest::topic_name()).expect("Invalid topic");
+
+        let request_topic_update =
+            Topic::new(SoftwareUpdateRequest::topic_name()).expect("Invalid topic");
+
+        let response_topic_list =
+            Topic::new(SoftwareListResponse::topic_name()).expect("Invalid topic");
+
+        let response_topic_update =
+            Topic::new(SoftwareUpdateResponse::topic_name()).expect("Invalid topic");
+
+        let errors_topic = Topic::new("tedge/errors").expect("Invalid topic");
+
+        let mqtt_client_config = mqtt_client::Config::default().with_packet_size(50 * 1024);
+
+        Self {
+            request_topics,
+            request_topic_list,
+            request_topic_update,
+            response_topic_list,
+            response_topic_update,
+            errors_topic,
+            mqtt_client_config,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SmAgent {
+    config: SmAgentConfig,
+    name: String,
+    user_manager: UserManager,
+    config_location: TEdgeConfigLocation,
+    persistance_store: AgentStateRepository,
+}
+
+impl SmAgent {
+    pub fn new(
+        name: &str,
+        user_manager: UserManager,
+        config_location: TEdgeConfigLocation,
+    ) -> Self {
+        let persistance_store = AgentStateRepository::new(&config_location);
+
+        Self {
+            config: SmAgentConfig::default(),
+            name: name.into(),
+            user_manager,
+            config_location,
+            persistance_store,
+        }
+    }
+
+    pub async fn start(&self) -> Result<(), AgentError> {
+        info!("Starting sm-agent");
+
+        let plugins = Arc::new(ExternalPlugins::open("/etc/tedge/sm-plugins")?);
+        if plugins.empty() {
+            error!("Couldn't load plugins from /etc/tedge/sm-plugins");
+            return Err(AgentError::NoPlugins);
+        }
+
+        let mqtt = Client::connect(self.name.as_str(), &self.config.mqtt_client_config).await?;
+        let mut errors = mqtt.subscribe_errors();
+        tokio::spawn(async move {
+            while let Some(error) = errors.next().await {
+                error!("{}", error);
+            }
+        });
+
+        let () = self.check_state_store(&mqtt).await?;
+
+        // * Maybe it would be nice if mapper/registry responds
+        let () = publish_capabilities(&mqtt).await?;
+
+        let () = self.subscribe_and_process(&mqtt, &plugins).await?;
+
+        Ok(())
+    }
+
+    async fn subscribe_and_process(
+        &self,
+        mqtt: &Client,
+        plugins: &Arc<ExternalPlugins>,
+    ) -> Result<(), AgentError> {
+        let mut operations = mqtt.subscribe(self.config.request_topics.clone()).await?;
+        while let Some(message) = operations.next().await {
+            info!("Request {:?}", message);
```

Comment:
This makes the agent a bit too verbose. It's good to log the operations (notably the updates), but not the raw messages.

Reply:
That should be debug level I suppose.